service/policy/db/attributes.go

```diff
@@ -55,227 +54,85 @@ func attributesValuesProtojson(valuesJSON []byte, attrFqn sql.NullString) ([]*po
-	if shouldGetValues {
-		fields = append(fields, &valuesOrder, &valuesJSON)
-	}
-	if opts.withKeyAccessGrants {
-		fields = append(fields, &grants)
-	}
-	if opts.withFqn {
-		fields = append(fields, &fqn)
-	}
-
-	err := row.Scan(fields...)
-	if err != nil {
-		return nil, db.WrapIfKnownInvalidQueryErr(err)
-	}
-
-	m := &common.Metadata{}
-	if metadataJSON != nil {
-		if err := protojson.Unmarshal(metadataJSON, m); err != nil {
+func hydrateAttribute(row *attributeQueryRow, logger *logger.Logger) (*policy.Attribute, error) {
+	metadata := &common.Metadata{}
+	if row.metadataJSON != nil {
+		if err := protojson.Unmarshal(row.metadataJSON, metadata); err != nil {
 			logger.Error("could not unmarshal metadata", slog.String("error", err.Error()))
 			return nil, err
 		}
 	}
-	var v []*policy.Value
-	if valuesJSON != nil {
-		v, err = attributesValuesProtojson(valuesJSON, fqn)
+
+	var values []*policy.Value
+	if row.valuesJSON != nil {
+		v, err := attributesValuesProtojson(row.valuesJSON, row.fqn)
 		if err != nil {
 			logger.Error("could not unmarshal values", slog.String("error", err.Error()))
 			return nil, err
 		}
+		values = v
 	}
-	var k []*policy.KeyAccessServer
-	if grants != nil {
-		k, err = db.KeyAccessServerProtoJSON(grants)
+
+	var grants []*policy.KeyAccessServer
+	if row.grants != nil {
+		k, err := db.KeyAccessServerProtoJSON(row.grants)
 		if err != nil {
 			logger.Error("could not unmarshal key access grants", slog.String("error", err.Error()))
 			return nil, err
 		}
+		grants = k
 	}
 
 	ns := &policy.Namespace{
-		Id:   namespaceID,
-		Name: namespaceName,
-		Fqn:  fmt.Sprintf("https://%s", namespaceName),
+		Id:   row.namespaceID,
+		Name: row.namespaceName,
+		Fqn:  fmt.Sprintf("https://%s", row.namespaceName),
 	}
 
 	attr := &policy.Attribute{
-		Id:        id,
-		Name:      name,
-		Rule:      attributesRuleTypeEnumTransformOut(rule),
-		Active:    &wrapperspb.BoolValue{Value: active},
-		Metadata:  m,
+		Id:        row.id,
+		Name:      row.name,
+		Rule:      attributesRuleTypeEnumTransformOut(row.rule),
+		Active:    &wrapperspb.BoolValue{Value: row.active},
+		Metadata:  metadata,
 		Namespace: ns,
-		Grants:    k,
-		Fqn:       fqn.String,
+		Grants:    grants,
+		Fqn:       row.fqn.String,
 	}
 
 	// Deactivations of individual values can unsync the order in the values_order column and the selected number of attribute values.
 	// In Go, int value 0 is not equal to 0, so check if they're not equal and more than 0 to check a potential count mismatch.
-	mismatchedCount := len(v) > 0 && len(valuesOrder) > 0 && len(valuesOrder) != len(v)
+	mismatchedCount := len(values) > 0 && len(row.valuesOrder) > 0 && len(row.valuesOrder) != len(values)
 
 	// sort the values according to the order
 	ordered := make([]*policy.Value, 0)
-	for _, order := range valuesOrder {
-		for _, value := range v {
+	for _, order := range row.valuesOrder {
```

Comment:
We should be able to do this in SQL more easily now like this:
https://github.com/opentdf/platform/blob/main/service/policy/db/query.sql#L206-L207

Reply:
This is now done in sql with `ORDER BY array_position` as shown here: https://github.com/opentdf/platform/pull/1450/files#diff-92033a056a1e64fa5b759cf500f2644f7e69e51fb26fa231dbe947bac600bf3fR357